mobile/mobile.go

```diff
@@ -181,51 +181,56 @@ func ConnectServer(nameSpace string, mailboxServer string, isDevServer bool,
 		return fmt.Errorf("unknown namespace: %s", nameSpace)
 	}
 
-	statusChecker, lndConnect, err := core.MailboxRPCConnection(
-		mailboxServer, pairingPhrase, localPriv, remotePub,
-		func(key *btcec.PublicKey) error {
-			mc.remoteKeyReceiveCallback.SendResult(
-				hex.EncodeToString(key.SerializeCompressed()),
-			)
-
-			return nil
-		}, func(data []byte) error {
-			parts := strings.Split(string(data), ": ")
-			if len(parts) != 2 || parts[0] != "Macaroon" {
-				return fmt.Errorf("authdata does " +
-					"not contain a macaroon")
-			}
-
-			macBytes, err := hex.DecodeString(parts[1])
-			if err != nil {
-				return err
-			}
-
-			mac := &macaroon.Macaroon{}
-			err = mac.UnmarshalBinary(macBytes)
-			if err != nil {
-				return fmt.Errorf("unable to decode "+
-					"macaroon: %v", err)
-			}
-
-			mc.mac = mac
-
-			mc.authDataCallback.SendResult(string(data))
+	// Since the connection function is blocking, we need to spin it off
+	// in another goroutine here. See https://pkg.go.dev/syscall/js#FuncOf.
+	go func() {
```

Comment:
With the `defer mMutex.Unlock()` above the lock on the map is released as soon as this goroutine is created. So it's possible the lock is released before the goroutine is actually running. Is there a possibility that the `mc` variable is accessed in another function (for the same `nameSpace`)? Then that would lead to a data race. If you want to avoid that you need another mutex on the `mc` variable itself (in addition to the one guarding the `m` map) to make sure no two calls can mutate the content (e.g. `mc.mac` or `mc.lndConn`) at the same time.

Reply:
mm, there is a possiblilty calls `ConnectServer` twice or `Status` while this is goroutine is running and triggers this. But when we're connecting we typically poll on `IsConnected` to find when the connection has succeeded.

Need to test if everything would still work with a second mutex there.